lib/doh.c

```diff
@@ -894,8 +1014,239 @@ UNITTEST void de_cleanup(struct dohentry *d)
+  for(i = 0; i < d->numhttps_rrs; i++)
+    free(d->https_rrs[i].val);
+#endif
 }
 
+#ifdef USE_HTTPSRR
+
+/*
+ * @brief decode the DNS name in a binary RRData
+ * @param buf points to the buffer (in/out)
+ * @param remaining points to the remaining buffer length (in/out)
+ * @param dnsname returns the string form name on success
+ * @return is 1 for success, error otherwise
+ *
+ * The encoding here is defined in
+ * https://tools.ietf.org/html/rfc1035#section-3.1
+ *
+ * The input buffer pointer will be modified so it points to
+ * just after the end of the DNS name encoding on output. (And
+ * that's why it's an "unsigned char **" :-)
+ */
+static CURLcode local_decode_rdata_name(unsigned char **buf, size_t *remaining,
+                                        char **dnsname)
+{
+  unsigned char *cp = NULL;
+  size_t rem = 0;
+  char *thename = NULL, *tp = NULL;
+  unsigned char clen = 0; /* chunk len */
+
+  if(!buf || !remaining || !dnsname)
+    return CURLE_OUT_OF_MEMORY;
+  rem = *remaining;
+  thename = calloc(1, CURL_MAXLEN_host_name);
+  if(!thename)
+    return CURLE_OUT_OF_MEMORY;
+  cp = *buf;
+  tp = thename;
+  clen = *cp++;
+  if(clen == 0) {
+    /* special case - return "." as name */
+    thename[0] = '.';
+    thename[1] = 0x00;
+  }
+  while(clen) {
+    if(clen > rem) {
+      free(thename);
+      return CURLE_OUT_OF_MEMORY;
+    }
+    if(((tp - thename) + clen) > CURL_MAXLEN_host_name) {
+      free(thename);
+      return CURLE_OUT_OF_MEMORY;
+    }
+    memcpy(tp, cp, clen);
+    tp += clen;
+    *tp++ = '.';
+    cp += clen;
+    rem -= (clen + 1);
+    clen = *cp++;
+  }
+  *buf = cp;
+  *remaining = rem - 1;
+  *dnsname = thename;
+  return CURLE_OK;
+}
+
+static CURLcode local_decode_rdata_alpn(unsigned char *rrval, size_t len,
+                                        char **alpns)
+{
+  /*
+   * spec here is as per draft-ietf-dnsop-svcb-https, section-7.1.1
+   * encoding is catenated list of strings each preceded by a one
+   * octet length
+   * output is comma-sep list of the strings
+   * implementations may or may not handle quoting of comma within
+   * string values, so we might see a comma within the wire format
+   * version of a string, in which case we'll precede that by a
+   * backslash - same goes for a backslash character
+   */
+  int remaining = (int) len;
+  char *oval, *op;
+  unsigned char *cp = rrval;
+
+  if(!alpns)
+    return CURLE_OUT_OF_MEMORY;
+  /* allow space to quote possibly every single character :-) */
+  oval = calloc(1, 2 * len);
+  if(!oval)
+    return CURLE_OUT_OF_MEMORY;
+  op = oval;
+  while(remaining > 0) {
+    size_t tlen = (size_t) *cp++;
+    remaining--;
+    if(tlen > (size_t)remaining || tlen > (size_t)((2 * len) - (op - oval)))
+      goto err;
+    memcpy(op, cp, tlen);
+    op += tlen;
+    if((size_t)(op - oval) >= ((size_t)2 * len))
+      goto err;
+    *op++ = COMMA_CHAR;
+    cp += tlen;
+    remaining -= (int)tlen;
+  }
+  oval[strlen(oval) - 1] = '\0';
```

Comment:
Can I propose that the above loop and decode is done into a dynbuf instead of this custom calloc + memcpy logic? dynbufs generally make it harder to mess up the memory fiddling.

Reply:
sounds right, but I'll revisit later, and there's still an escaping TODO as well, so leaving this unresolved for now.

Reply:
took a stab at that in https://github.com/curl/curl/pull/11922/commits/ecf5952a8b668d59f7da4d082c91d555f3689aec - could be better if ``dynbuf`` percolates upwards in the data structure but this seems at least to work:-)